services/configuration_repository/configuration_repository_service.rb

```diff
@@ -0,0 +1,252 @@
+              full_name: "Clone User credentials"
+            )
+          ]
+        )
+      ]
+
+      users.each do |user|
+        # Fist need to serialize the provider credentials and ignore the `ci_user`
+        # instance variable. The reasoning is since if you serialize the `user`
+        # first, you will call `to_object_map` on the `ci_user`, which holds
+        # reference to a user. This will go on indefinitely
+        user.provider_credentials.map! do |credential|
+          credential.to_object_dictionary(ignore_instance_variables: [:@ci_user])
+        end
+      end
+
+      JSON.pretty_generate(users.map(&:to_object_dictionary))
+    end
+
+    # Creates an empty json array file in the configuration repository
+    #
+    # @param  [String] file_path
+    def create_remote_json_file(file_path, json_string: "[]")
+      not_implemented(__method__)
+    end
+
+    #####################################################
+    # @!group Boolean Helpers
+    #####################################################
+
+    # Configuration will fail if the `file_path` file contents are not a
+    # JSON array
+    #
+    # @param  [String] file_path
+    # @return [Boolean]
+    def remote_file_a_json_array?(file_path)
+      not_implemented(__method__)
+    end
+
+    # This method starts the watching of the ci-config repo folder changes
+    # so we can report every file being changed in a real-time basis.
+    # @param [Git::Base] git
+    # @param [String] path, the path of the repo which changes are being watched.
+    def watch_for_changes!(git, path)
+      require "ruby-watchman"
+      require "socket"
+      sockname = RubyWatchman.load(
+        `watchman --output-encoding=bser get-sockname`
+      )["sockname"]
+      raise unless $?.exitstatus.zero?
+
+      unwatch_changes unless @socket.nil?
+
+      @socket = UNIXSocket.open(sockname) do |socket|
+        root = Pathname.new(path).realpath.to_s
+        roots = RubyWatchman.query(["watch-list"], socket)["roots"]
+        unless roots.include?(root)
+          # this path isn't being watched yet; try to set up watch
+          result = RubyWatchman.query(["watch", root], socket)
+
+          # root_restrict_files setting may prevent Watchman from working
+          raise if result.key?("error")
+        end
+
+        query = ["query", root, {
+        "expression" => ["anyof",
+                         ["match", "**/*.json", "wholename", { "includedotfiles" => false }]],
+          "fields" => ["name"]
+        }]
+        paths = RubyWatchman.query(query, socket)
+
+        # could return error if watch is removed
+        raise if paths.key?("error")
+
+        unless ConfigurationRepositoryService.mutex.locked?
+          ConfigurationRepositoryService.mutex.synchronize do
+            handle_repo_changes(git, paths["files"])
+          end
+        end
+      end
+    end
+
+    def unwatch_changes
+      `watchman shutdown-server`
+      @socket&.close
```

Comment:
why the `&`?

Reply:
Safety purposes, `@socket` might be nil at some point, I don't want to crash because of the reference to the socket being lost.

Reply:
Ha, that's fine, I'm just not familiar with that Ruby language feature, any idea when it was added?

Reply:
2.3 😄 http://mitrev.net/ruby/2015/11/13/the-operator-in-ruby/